source/opt/merge_return_pass.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "merge_return_pass.h"
+
+#include "instruction.h"
+#include "ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status MergeReturnPass::Process(ir::IRContext *irContext) {
+  InitializeProcessing(irContext);
+
+  context()->InvalidateAnalysesExceptFor(ir::IRContext::kAnalysisDefUse);
+
+  bool modified = false;
+  for (auto &function : *get_module()) {
+    std::vector<ir::BasicBlock *> returnBlocks = collectReturnBlocks(&function);
+    modified |= mergeReturnBlocks(&function, returnBlocks);
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+std::vector<ir::BasicBlock *> MergeReturnPass::collectReturnBlocks(
+    ir::Function *function) {
+  std::vector<ir::BasicBlock *> returnBlocks;
+  for (auto &block : *function) {
+    ir::Instruction &terminator = *block.tail();
+    if (terminator.opcode() == SpvOpReturn ||
+        terminator.opcode() == SpvOpReturnValue) {
+      returnBlocks.push_back(&block);
+    }
+  }
+
+  return std::move(returnBlocks);
+}
+
+bool MergeReturnPass::mergeReturnBlocks(
+    ir::Function *function, const std::vector<ir::BasicBlock *> &returnBlocks) {
+  if (returnBlocks.size() <= 1) {
+    // No work to do.
+    return false;
+  }
+
+  // Create a label for the new return block
+  std::unique_ptr<ir::Instruction> returnLabel(
+      new ir::Instruction(SpvOpLabel, 0u, TakeNextId(), {}));
+  uint32_t returnId = returnLabel->result_id();
+
+  // Create the new basic block
+  std::unique_ptr<ir::BasicBlock> returnBlockUPtr(
+      new ir::BasicBlock(std::move(returnLabel)));
+  function->AddBasicBlock(std::move(returnBlockUPtr));
+  ir::BasicBlock &returnBlock = *(--function->end());
+
+  // Create the PHI for the merged block (if necessary)
+  // Create new return
+  std::vector<ir::Operand> phiOps;
+  for (auto block : returnBlocks) {
+    if (block->tail()->opcode() == SpvOpReturnValue) {
+      phiOps.push_back(
+          {SPV_OPERAND_TYPE_ID, {block->tail()->GetSingleWordInOperand(0u)}});
+      phiOps.push_back({SPV_OPERAND_TYPE_ID, {block->id()}});
+    }
+  }
+
+  if (!phiOps.empty()) {
+    // Need a PHI node to select the correct return value.
+    uint32_t phiResultId = TakeNextId();
+    uint32_t phiTypeId = function->type_id();
```

Comment:
This is a sad irregularity of SPIR-V.   It's the *ONE* place where the type of the thing is not the same as the typeid of the thing.  :-(

Reply:
I'm attempting to get the return type of the function here. Function::type_id() returns the type id of the defining instruction (OpFunction). Is this not correct?

Reply:
Your code is correct.   It's just an irregularity of the spec.    The result id of the OpFunction is an OpFunction.  But the type id is the type of the value returned by the function, not type of the function.
Never mind. It just bugs me.